### PR TITLE
[Bugfix] Handle `best_of>1` case by disabling speculation.

### PR DESCRIPTION
**Pull request by tdoublep**
This PR solves #6137 by disabling speculation for batches that contain any request with `best_of>1`.

This approach ensures that we can handle requests with`best_of>1` without failure, but may have a downside that a single user sending requests with `best_of>1` can potentially ruin performance for other users with `best_of=1`. 

An alternative solution could just be to raise on those individual requests and give a message to the user like `best_of > 1 is not supported when speculative decoding is enabled`. 

I can also implement that if preferred. What do you think @cadedaniel @njhill  ?

### CODE REVIEW

**Comment by cadedaniel** [comment]
Thanks for the fix -- approach looks good to me.

On whether or not we should support this -- for performance we would want to disable this feature or support it natively in spec decode. I am fine having this in, can we log once if this happens so there's a hint of the performance degredation to users?

File: vllm/engine/output_processor/multi_step.py [hunk 1]
```diff
@@ -72,6 +79,11 @@ def process_outputs(self, sequence_group: SequenceGroup,
         seqs = sequence_group.get_seqs(status=SequenceStatus.RUNNING)
 
         assert seqs, "expected running sequences"
+
+        if len(seqs) > 1 or sequence_group.sampling_params.best_of > 1:
```
Comment (cadedaniel) [inline vllm/engine/output_processor/multi_step.py:83]: add a comment here explaining what's going on. also can we explicitly fail if beam search is enabled?

File: vllm/spec_decode/spec_decode_worker.py [hunk 2]
```diff
@@ -352,10 +352,15 @@ def _should_disable_all_speculation(
...
-        disable_all_speculation = (execute_model_req.running_queue_size >=
-                                   self.disable_by_batch_size)
+        if execute_model_req.running_queue_size >= self.disable_by_batch_size:
+            return True
 
-        return disable_all_speculation
+        # Disable speculation if best_of > 1 for any sequence group
+        for seq_group_metadata in execute_model_req.seq_group_metadata_list:
+            if seq_group_metadata.sampling_params.best_of > 1:
+                return True
```
Comment (cadedaniel) [inline vllm/spec_decode/spec_decode_worker.py:361]: suggest moving this to a method so you can add a docstring with more explanation on how this works (e.g. performance tradeoff)

File: vllm/engine/output_processor/multi_step.py (hunk 1 above)
Comment (tdoublep) [reply vllm/engine/output_processor/multi_step.py:86]: done. 

re: failing on beam search, I think this is best done *before* we put the requests in the batch so I've added some code in `add_request` to that effect.

File: vllm/spec_decode/spec_decode_worker.py (hunk 2 above)
Comment (tdoublep) [reply vllm/spec_decode/spec_decode_worker.py]: done

**Comment by tdoublep** [comment]
> I am fine having this in, can we log once if this happens so there's a hint of the performance degredation to users?

I added a warning when we disable speculation due to n>1 or best_of>1.

**Pull request closed without merging**